### PR TITLE
Fluentd

### PR DESCRIPTION
**Pull request by mzaneri**
* Logging plugin using Fluentd as the unified logging layer

* Fluentd is implemented as a Daemonset, reading logs from all containers on the node. By default it sends all output to stdout. 

* Fluentd options can be edited in fluentd.conf to collect logs from various sources, filter and tag according to rules, and send to various aggregators. This can be edited via Kustomize.

* Optional Elasticsearch and Kibana integration for complete EFK logging stack and visualization.

* Elasticsearch requires init containers to handle permissions, and 1gb of memory for stability. Default storage is set to 20GB.

* Elasticsearch using the existing rook/ceph setups to handle the persistent volume claims.

* There is a different fluentd.conf when using Elasticsearch and Kibana.

* Elasticsearch is implemented as a StatefulSet

* Add Fluentd to installer, no ES/Kibana as default option, flag for full EFK stack

* Unit tests and api tests included and passing.

### CODE REVIEW

File: scripts/Manifest [hunk 1]
```diff
@@ -1,4 +1,4 @@
-KUBERNETES_VERSION=1.15.3
+KUBERNETES_VERSION=1.15.2
```
Comment (emosbaugh) [inline scripts/Manifest]: ?

File: web/src/test/installers.ts [hunk 2]
```diff
@@ -492,4 +521,32 @@ spec:
       expect(i.flags()).to.equal(`velero-namespace=velero`);
     });
   });
+        
+  describe("fluentd minimum spec flags", () => {
+    it("should not generate any flags", () => {
+
+        const i = Installer.parse(fluentdMin);
```
Comment (emosbaugh) [inline web/src/test/installers.ts]: indentation

Comment (mzaneri) [reply web/src/test/installers.ts]: this is a dupe test actually ill delete it

File: scripts/Manifest (hunk 1 above)
Comment (mzaneri) [reply scripts/Manifest]: done

File: addons/fluentd/1.3/install.sh [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        kubectl apply -k "$logging_dst/"
+        kubectl apply -k "$fluentd_dst/"
+    fi
+
+}
+
+fluentd_outro() {
+    printf "\n"
+    printf "\n"
+    printf "The UI of Kibana has been exposed on NodePort ${GREEN}30887${NC}\n"
```
[the remark this answers is not in the document]
Comment (mzaneri) [reply addons/fluentd/1.3/install.sh]: edited

File: addons/fluentd/1.3/Manifest [hunk 4]
```diff
@@ -0,0 +1,5 @@
+image fluentd fluent/fluentd-kubernetes-daemonset:v1.7.4-debian-elasticsearch7-1.0
+image elasticsearch docker.elastic.co/elasticsearch/elasticsearch:7.5.0
+image kibana docker.elastic.co/kibana/kibana:7.5.0
+image busybox busybox
```
[the remark this answers is not in the document]
Comment (mzaneri) [reply addons/fluentd/1.3/Manifest]: edited

File: web/src/client/index_test.ts [hunk 5]
```diff
@@ -156,6 +165,14 @@ describe("POST /installer", () => {
   });
 
   describe("kots", () => {
+    it(`should return 201 "https://kurl.sh/4723751"`, async () => {
```
[the remark this answers is not in the document]
Comment (mzaneri) [reply web/src/client/index_test.ts]: changes

File: addons/fluentd/1.3/elasticsearch/elasticsearch-statefulset.yaml [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+          mountPath: /usr/share/elasticsearch/data
+        env:
+          - name: cluster.name
+            value: k8s-logs
+          - name: NODE_NAME
+            valueFrom:
+              fieldRef:
+                fieldPath: metadata.name
+          - name: node.name
+            value: "$(NODE_NAME).elasticsearch"
```
[the remark this answers is not in the document]
Comment (mzaneri) [reply addons/fluentd/1.3/elasticsearch/elasticsearch-statefulset.yaml]: discussed

File: addons/fluentd/1.3/elasticsearch/elasticsearch-statefulset.yaml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+          - name: discovery.seed_hosts
+            value: "es-cluster-0.elasticsearch,es-cluster-1.elasticsearch,es-cluster-2.elasticsearch"
+          - name: cluster.initial_master_nodes
+            value: "es-cluster-0.elasticsearch,es-cluster-1.elasticsearch,es-cluster-2.elasticsearch"
+          - name: ES_JAVA_OPTS
+            value: "-Xms1024m -Xmx1024m"
+      initContainers:
+      - name: fix-permissions
```
[the remark this answers is not in the document]
Comment (mzaneri) [reply addons/fluentd/1.3/elasticsearch/elasticsearch-statefulset.yaml]: Did not get this to work- with a pod security context

File: web/src/client/index_test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (mzaneri) [reply web/src/client/index_test.ts]: fixed, passes tests

**Pull request merged**